internal/config/variables/variables.go

```diff
@@ -0,0 +1,531 @@
+					"%q, place a variable definition block in your waypoint.hcl file.",
+					pbv.Name, pbv.Name),
+			})
+			continue
+		}
+
+		var source string
+		switch pbv.Source.(type) {
+		case *pb.Variable_Cli:
+			source = sourceCLI
+		case *pb.Variable_File_:
+			source = sourceFile
+		case *pb.Variable_Env:
+			source = sourceEnv
+		case *pb.Variable_Vcs:
+			source = sourceVCS
+		case *pb.Variable_Server:
+			source = sourceServer
+		}
+
+		// We have to specify the three different simple types we support -- string,
+		// bool, number -- when doing the below evaluation of hcl expressions
+		// because of our translation to-and-from protobuf format.
+		// While cty allows us to parse all simple types as LiteralValueExpr, we
+		// have to first translate the pb values back into cty values, thus
+		// necessitating a separate case statement for each simple type
+		var expr hclsyntax.Expression
+		switch pbv.Value.(type) {
+
+		case *pb.Variable_Hcl:
+			value := pbv.Value.(*pb.Variable_Hcl).Hcl
+			fakeFilename := fmt.Sprintf("<value for var.%s from server>", pbv.Name)
+			expr, diags = hclsyntax.ParseExpression([]byte(value), fakeFilename, hcl.Pos{Line: 1, Column: 1})
+		case *pb.Variable_Str:
+			value := pbv.Value.(*pb.Variable_Str).Str
+			expr = &hclsyntax.LiteralValueExpr{Val: cty.StringVal(value)}
+		case *pb.Variable_Bool:
+			value := pbv.Value.(*pb.Variable_Bool).Bool
+			expr = &hclsyntax.LiteralValueExpr{Val: cty.BoolVal(value)}
+		case *pb.Variable_Num:
+			value := pbv.Value.(*pb.Variable_Num).Num
+			expr = &hclsyntax.LiteralValueExpr{Val: cty.NumberIntVal(value)}
+		}
+
+		val, valDiags := expr.Value(nil)
+		if valDiags.HasErrors() {
+			diags = append(diags, valDiags...)
+			return nil, diags
+		}
+
+		if variable.Type != cty.NilType {
+			var err error
+			val, err = convert.Convert(val, variable.Type)
+			if err != nil {
+				diags = append(diags, &hcl.Diagnostic{
+					Severity: hcl.DiagError,
+					Summary:  "Invalid value for variable",
+					Detail: fmt.Sprintf(
+						"The value set for variable %q from source %q is not compatible with the variable's type constraint: %s.",
+						pbv.Name,
+						source,
+						err,
+					),
+				})
+				val = cty.DynamicVal
+			}
+		}
+
+		iv[pbv.Name] = &InputValue{
+			Source: source,
+			Value:  val,
+			Expr:   expr,
+		}
+	}
+
+	// check that all variables have a set value, including default of null
+	for name := range vs {
+		_, found := iv[name]
+		if !found {
+			return nil, append(diags, &hcl.Diagnostic{
+				Severity: hcl.DiagError,
+				Summary:  fmt.Sprintf("Unset variable %q", name),
+				// TODO krantzinator: add our docs link here
+				Detail: "A variable must be set or have a default value; see " +
+					"[docs] for " +
+					"details.",
+			})
+		}
+	}
+
+	return iv, diags
+}
+
+// LoadVCSFiles loads any *.auto.wpvars(.json) files in the source repo
+func LoadVCSFiles(wd string) ([]*pb.Variable, hcl.Diagnostics) {
```

Comment:
Naming bikeshedding again, I'd drop VCS from this. The variables package itself doesn't care if this is a VCS repo or not. It's just loading files from a working directory. I'd rename this to like `LoadAutoFiles` or something.

Reply:
Makes sense, done!